### PR TITLE
fix: edit load test missing the userclasses data

### PR DESCRIPTION
**Pull request by alterhu2020**
fix the problem userclasses missing issue when edit the load test

### CODE REVIEW

**Comment by cyberw** [comment]
Does that work though? If you change the selected Users, that wont affect the ones that are already running, right?

**Comment by alterhu2020** [comment]
> Does that work though? If you change the selected Users, that wont affect the ones that are already running, right?

Yes,I had tested from my machine, It works as expected. It only reused the selected user we had picked previously.

**Comment by cyberw** [comment]
Nice! Can you add a unit test that validates it?

Here’s some inspiration: https://github.com/locustio/locust/blob/4989f2a46a5cb8057f5461b5baf336a4243aa846/locust/test/test_web.py#L410

File: locust/web.py [hunk 1]
```diff
@@ -164,7 +164,8 @@ def swarm() -> Response:
                             user_classes[user_class_name] = user_class_object
 
                 else:
-                    user_classes = self.environment.available_user_classes
+                    user_classes = {key: value for (key, value) in self.environment.available_user_classes.items if
```
Comment (mikenester) [inline locust/web.py:167]: I tried out this change and got
```
File "/Users/michaelnester/git/locust/locust/web.py", line 167, in swarm
  user_classes = {
TypeError: 'builtin_function_or_method' object is not iterable
```

I think it needs to be `self.environment.available_user_classes.items()`

**Comment by mikenester** [comment]
This change breaks `test_swarm_defaults_to_all_available_userclasses_when_userclass_picker_is_active_and_no_userclass_in_payload`.

The workflow is:

1. Run Locust w/ class picker and multiple user classes to choose from
2. Select one user class
  - More specifically: select up to **N - 1** user classes where N is the total number of available user classes  
3. Run the test
4. Stop the test and start a new test, but don't select any user classes
  - This defaults to all available user classes

Only the user class from the first run will be used

**Comment by mikenester** [comment]
@alterhu2020 I think i got a fix for this: https://github.com/locustio/locust/compare/master...mikenester:locust:fix-class-picker-bug-edit-running-test

File: locust/web.py (hunk 1 above)
Comment (alterhu2020) [reply locust/web.py:167]: yes, sorry. My typo error. I will put the unit test code soon

**Comment by alterhu2020** [comment]
> @alterhu2020 I think i got a fix for this: [master...mikenester:locust:fix-class-picker-bug-edit-running-test](https://github.com/locustio/locust/compare/master...mikenester:locust:fix-class-picker-bug-edit-running-test)

Yes, the code is better than mine, please go ahead merged it. thanks

**Comment by mikenester** [comment]
@alterhu2020 If you merge this https://github.com/alterhu2020/locust/pull/1, it **should** automatically update this PR and run the tests

**Comment by alterhu2020** [comment]
BTW， @mikenester 
When I run the locust  unit test code or run the `main.py` or `_main_.py` from pycharm directly, it often returns following error:
```

Error
Traceback (most recent call last):
  File "C:\Users\Administrator\.conda\envs\locust\lib\unittest\loader.py", line 34, in testFailure
    raise self._exception
ImportError: Failed to import test module: locust
Traceback (most recent call last):
  File "C:\Users\Administrator\.conda\envs\locust\lib\unittest\loader.py", line 154, in loadTestsFromName
    module = __import__(module_name)
  File "D:\locust\locust\__init__.py", line 18, in <module>
    from .contrib.fasthttp import FastHttpUser
  File "D:\locust\locust\contrib\fasthttp.py", line 24, in <module>
    from locust.env import Environment
  File "D:\locust\locust\env.py", line 18, in <module>
    from .web import WebUI
  File "D:\locust\locust\web.py", line 6, in <module>
    from html import escape
  File "D:\locust\locust\html.py", line 6, in <module>
    from .stats import sort_stats
ImportError: attempted relative import with no known parent package

```

I know it related that we use the relative import path, but I don't know how to bypass it? Have we any guideline that can help people  quick to setup the develop contribution environment? thanks

**Comment by alterhu2020** [comment]
> @alterhu2020 If you merge this [alterhu2020#1](https://github.com/alterhu2020/locust/pull/1), it **should** automatically update this PR and run the tests

LGTM, Done as well

**Comment by cyberw** [comment]
Nice! But the test doesnt actually validate that the Users changed, does it?

**Comment by mikenester** [comment]
> Nice! But the test doesnt actually validate that the Users changed, does it?

Perhaps I misunderstood, but I thought we **don't** want the users to change. So an example of the expected workflow should be:

1. Run locust with `--class-picker` and available user classes: **UserClass1**, **UserClass2**, **UserClass3**
2. Run a test with just **UserClass1**
3. Edit the running test. User Class Count & Spawn Rate values can be arbitrary
4. Validate that only **UserClass1** is active
    - The reported bug being that instead, **UserClass1**, **UserClass2** & **UserClass3** were all active
    - Validated on line 440 of `locust/test/test_web.py` in this PR

**Comment by alterhu2020** [comment]
> > Nice! But the test doesnt actually validate that the Users changed, does it?
> 
> Perhaps I misunderstood, but I thought we **don't** want the users to change. So an example of the expected workflow should be:
> 
> 1. Run locust with `--class-picker` and available user classes: **UserClass1**, **UserClass2**, **UserClass3**
> 2. Run a test with just **UserClass1**
> 3. Edit the running test. User Class Count & Spawn Rate values can be arbitrary
> 4. Validate that only **UserClass1** is active
>    
>    * The reported bug being that instead, **UserClass1**, **UserClass2** & **UserClass3** were all active
>    * Validated on line 440 of `locust/test/test_web.py` in this PR

Yes. I think @mikenester is right.  @cyberw for the Users Changed scenario I think It related to another feature I raised in issue: https://github.com/locustio/locust/issues/2169, right? If we can support that I think it's a great improvement. 
Thanks Team.

**Comment by mikenester** [comment]
> BTW， @mikenester When I run the locust unit test code or run the `main.py` or `_main_.py` from pycharm directly, it often returns following error:
> 
> ```
> 
> Error
> Traceback (most recent call last):
>   File "C:\Users\Administrator\.conda\envs\locust\lib\unittest\loader.py", line 34, in testFailure
>     raise self._exception
> ImportError: Failed to import test module: locust
> Traceback (most recent call last):
>   File "C:\Users\Administrator\.conda\envs\locust\lib\unittest\loader.py", line 154, in loadTestsFromName
>     module = __import__(module_name)
>   File "D:\locust\locust\__init__.py", line 18, in <module>
>     from .contrib.fasthttp import FastHttpUser
>   File "D:\locust\locust\contrib\fasthttp.py", line 24, in <module>
>     from locust.env import Environment
>   File "D:\locust\locust\env.py", line 18, in <module>
>     from .web import WebUI
>   File "D:\locust\locust\web.py", line 6, in <module>
>     from html import escape
>   File "D:\locust\locust\html.py", line 6, in <module>
>     from .stats import sort_stats
> ImportError: attempted relative import with no known parent package
> ```
> 
> I know it related that we use the relative import path, but I don't know how to bypass it? Have we any guideline that can help people quick to setup the develop contribution environment? thanks

Not sure if I can help with this one. I don't think you'll be able to successfully run individual files, since relative imports get in python can get broken in that scenario. 

In terms of testing, I personally modify `tox.ini` in the following way. Change line 30 from

`python3 -m coverage run -m unittest discover []`

to the intended unittest path:

`python3 -m coverage run -m unittest locust.tests.test_web.TestWebUI.test_exceptions_csv`

Then you can run:
- `tox -e py37` if you're using python 3.7
- `tox -e py38` if you're using python 3.8
- `tox -e py39` if you're using python 3.9
- `tox -e py310` if you're using python 3.10

**Comment by alterhu2020** [comment]
> > BTW， @mikenester When I run the locust unit test code or run the `main.py` or `_main_.py` from pycharm directly, it often returns following error:
> > ```
> > 
> > Error
> > Traceback (most recent call last):
> >   File "C:\Users\Administrator\.conda\envs\locust\lib\unittest\loader.py", line 34, in testFailure
> >     raise self._exception
> > ImportError: Failed to import test module: locust
> > Traceback (most recent call last):
> >   File "C:\Users\Administrator\.conda\envs\locust\lib\unittest\loader.py", line 154, in loadTestsFromName
> >     module = __import__(module_name)
> >   File "D:\locust\locust\__init__.py", line 18, in <module>
> >     from .contrib.fasthttp import FastHttpUser
> >   File "D:\locust\locust\contrib\fasthttp.py", line 24, in <module>
> >     from locust.env import Environment
> >   File "D:\locust\locust\env.py", line 18, in <module>
> >     from .web import WebUI
> >   File "D:\locust\locust\web.py", line 6, in <module>
> >     from html import escape
> >   File "D:\locust\locust\html.py", line 6, in <module>
> >     from .stats import sort_stats
> > ImportError: attempted relative import with no known parent package
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I know it related that we use the relative import path, but I don't know how to bypass it? Have we any guideline that can help people quick to setup the develop contribution environment? thanks
> 
> Not sure if I can help with this one. I don't think you'll be able to successfully run individual files, since relative import in python can get broken in that scenario.
> 
> In terms of testing, I personally modify `tox.ini` in the following way. Change line 30 from
> 
> `python3 -m coverage run -m unittest discover []`
> 
> to the intended unittest path:
> 
> `python3 -m coverage run -m unittest locust.tests.test_web.TestWebUI.test_exceptions_csv`
> 
> Then you can run:
> 
> * `tox -e py37` if you're using python 3.7
> * `tox -e py38` if you're using python 3.8
> * `tox -e py39` if you're using python 3.9
> * `tox -e py310` if you're using python 3.10

Thanks the tips, but how we can active the debugger, I want to read the code step by step...

**Comment by cyberw** [comment]
I misunderstood what you were trying to do :)

This looks good to me now. Should I merge?

**Comment by mikenester** [comment]
> This looks good to me now. Should I merge?

Yeah, I think it's good to go 👍

**Comment by cyberw** [comment]
thanks!

**Pull request merged**